context.go

```diff
@@ -130,79 +129,61 @@ var testContextData = struct {
 }{service: make(map[string][]byte, 0)}
 
 // Save takes an identifier and an interface. The interface will be network.Marshaled
-// and saved under a filename based on the identifier. An eventual error will be returned.
-// If contextDataPath is non-empty, the destination is a file: it will be created
-// with rw-r----- permissions (0640). If the file already exists, it will be overwritten.
-//
-// The path to the file is chosen as follows:
-//   Mac: ~/Library/Conode/Services
-//   Other Unix: ~/.local/share/conode
-//   Windows: $HOME$\AppData\Local\Conode
-// If the directory doesn't exist, it will be created using rwxr-x---
-// permissions (0750).
-//
-// The path can be overridden with the environmental variable "CONODE_SERVICE_PATH".
+// and saved in the database under the bucket named after the service name.
+// The database must be created by the server prior to using this function.
 func (c *Context) Save(id string, data interface{}) error {
 	buf, err := network.Marshal(data)
 	if err != nil {
 		return err
 	}
-	fname := c.absFilename(id)
-	if getContextDataPath() == "" {
-		testContextData.Lock()
-		testContextData.service[fname] = buf
-		testContextData.Unlock()
-		return nil
-	}
-
-	return ioutil.WriteFile(fname, buf, 0640)
+	bucketName := ServiceFactory.Name(c.ServiceID())
+	return c.manager.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucketName))
+		return b.Put([]byte(id), buf)
+	})
 }
 
 // Load takes an id and returns the network.Unmarshaled data. If an error
-// occurs, the data is nil. See Save() for where the files are saved.
-//
-// If no data is found, it returns an error.
+// occurs, the data is nil. If no data is found, it returns an error.
 func (c *Context) Load(id string) (interface{}, error) {
 	var buf []byte
-	if getContextDataPath() == "" {
-		var ok bool
-		testContextData.Lock()
-		buf, ok = testContextData.service[c.absFilename(id)]
-		testContextData.Unlock()
-		if !ok {
-			return nil, errors.New("this entry doesn't exist")
-		}
-	} else {
-		var err error
-		buf, err = ioutil.ReadFile(c.absFilename(id))
-		if err != nil {
-			return nil, err
-		}
+	bucketName := ServiceFactory.Name(c.ServiceID())
+	c.manager.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket([]byte(bucketName))
+		v := b.Get([]byte(id))
+		buf = make([]byte, len(v))
```

Comment:
Because of Boltdb's API, you are going to have this make and copy thing a lot. It is not really more readable, but it is more compact to do:
v:= b.Get([]byte(id))
return append([]byte(nil), v...)

It theoretically also informs the runtime that it could give you memory that has not been zeroed, because it knows for sure that it will be filling it before you get it back control. But I bet (a) there's no measurable perf impact from that, and (b) Go doesn't do it.

Reply:
I don't think this will work. The callback needs to return an error type.